### ISSUE TITLE
USB device still active even after physically unplugged

### ISSUE THREAD

**Issue by tchernomax** [systemd/systemd#7587]
### Submission type

 - Bug report

### systemd version the issue has been seen with

```
% systemctl --version
systemd 235
+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN default-hierarchy=hybrid
```

### Used distribution

> Archlinux

### In case of bug report: Expected behaviour you didn't see

Device plugged:
```
max@mde-oxalide % cat /etc/udev/rules.d/70-yubikey.rules
ACTION=="add", SUBSYSTEM=="usb", ENV{ID_VENDOR_ID}=="1050", ENV{ID_MODEL_ID}=="0010", TAG+="systemd", ENV{SYSTEMD_ALIAS}="/sys/subsystem/usb/yubikey"

max@mde-oxalide % sudo udevadm info -p '/devices/pci0000:00/0000:00:14.0/usb1/1-1'
P: /devices/pci0000:00/0000:00:14.0/usb1/1-1
…
max@mde-oxalide % systemctl status sys-subsystem-usb-yubikey.device
● sys-subsystem-usb-yubikey.device - Yubikey (v1 or v2)
   Loaded: loaded
   Active: active (plugged) since Fri 2017-12-08 15:09:32 CET; 10min ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1
```
When the device is unplugged (physically):
```
max@mde-oxalide % sudo udevadm info -p '/devices/pci0000:00/0000:00:14.0/usb1/1-1'
syspath not found

max@mde-oxalide % systemctl status sys-subsystem-usb-yubikey.device
● sys-subsystem-usb-yubikey.device - Yubikey (v1 or v2)
   Loaded: loaded
   Active: inactive (dead) …
```

### In case of bug report: Unexpected behaviour you saw

When the device is unplugged (physically):
```
max@mde-oxalide % sudo udevadm info -p '/devices/pci0000:00/0000:00:14.0/usb1/1-1'
syspath not found

max@mde-oxalide % systemctl status sys-subsystem-usb-yubikey.device
● sys-subsystem-usb-yubikey.device - Yubikey (v1 or v2)
   Loaded: loaded
   Active: active (plugged) since Fri 2017-12-08 15:09:32 CET; 10min ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1
```

### In case of bug report: Steps to reproduce the problem

Udev TAG+="systemd" an usb device an try to plug/unplug it.

### Additional information

When I unplug the device:
```
max@mde-oxalide % sudo udevadm monitor -ku --subsystem-match=usb
monitor will print the received events for:
UDEV - the event which udev sends out after rule processing
KERNEL - the kernel uevent

KERNEL[1117.750522] unbind   /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0 (usb)
KERNEL[1117.750623] remove   /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0 (usb)
KERNEL[1117.751051] unbind   /devices/pci0000:00/0000:00:14.0/usb1/1-1 (usb)
KERNEL[1117.751158] remove   /devices/pci0000:00/0000:00:14.0/usb1/1-1 (usb)
UDEV  [1117.753043] unbind   /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0 (usb)
UDEV  [1117.754020] remove   /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0 (usb)
UDEV  [1117.755406] unbind   /devices/pci0000:00/0000:00:14.0/usb1/1-1 (usb)
UDEV  [1117.771183] remove   /devices/pci0000:00/0000:00:14.0/usb1/1-1 (usb)
^C% 
```

It's exactly the same issue as #7018 but I use systemd 235, feel free to close this issue and reopen #7018 if you want.

**Comment by poettering** [comment]
does this work correctly if SYSTEMD_ALIAS is not in the mix?

**Comment by tchernomax** [comment]
@poettering no, same problem without the SYSTEMD_ALIAS.

```
max@mde-oxalide % sudo udevadm info -p '/devices/pci0000:00/0000:00:14.0/usb1/1-1'
P: /devices/pci0000:00/0000:00:14.0/usb1/1-1
…

max@mde-oxalide % systemctl status sys-devices-pci0000:00-0000:00:14.0-usb1-1\\x2d1.device
● sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device - Yubikey (v1 or v2)
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 18:54:11 CET; 2min 25s ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1
```
Device unplugged:
```
max@mde-oxalide % sudo udevadm info -p '/devices/pci0000:00/0000:00:14.0/usb1/1-1'
syspath not found
max@mde-oxalide % systemctl status sys-devices-pci0000:00-0000:00:14.0-usb1-1\\x2d1.device
● sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device - Yubikey (v1 or v2)
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 18:54:11 CET; 2min 32s ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1
```

**Comment by poettering** [comment]
that's really really strange... Do you see any interesting logging when you do this with debug logging on? (i.e. issue "systemd-analyze set-log-level debug" beforehand?)

**Comment by tchernomax** [comment]
@poettering Here is what I see when I plug the yubikey:
```
déc. 11 22:46:05 mde-oxalide kernel: usb 1-1: new low-speed USB device number 11 using xhci_hcd
déc. 11 22:46:06 mde-oxalide systemd[1]: systemd-udevd.service: Got notification message from PID 347 (WATCHDOG=1)
déc. 11 22:46:06 mde-oxalide kernel: input: Yubico Yubico Yubikey II as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/0003:1050:0010.0021/input/input61
déc. 11 22:46:06 mde-oxalide kernel: hid-generic 0003:1050:0010.0021: input,hidraw0: USB HID v1.11 Keyboard [Yubico Yubico Yubikey II] on usb-0000:00:14.0-1/input0
déc. 11 22:46:06 mde-oxalide mtp-probe[26312]: checking bus 1, device 11: "/sys/devices/pci0000:00/0000:00:14.0/usb1/1-1"
déc. 11 22:46:06 mde-oxalide mtp-probe[26312]: bus: 1, device: 11 was not an MTP device
déc. 11 22:46:06 mde-oxalide systemd[1]: dev-bus-usb-001-011.device: Changed dead -> plugged
déc. 11 22:46:06 mde-oxalide upowerd[698]: unhandled action 'bind' on /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/0003:1050:0010.0021
déc. 11 22:46:06 mde-oxalide upowerd[698]: unhandled action 'bind' on /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0
déc. 11 22:46:06 mde-oxalide upowerd[698]: unhandled action 'bind' on /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1
déc. 11 22:46:06 mde-oxalide systemd[1]: systemd-logind.service: Got notification message from PID 567 (FDSTORE=1, FDNAME=session-c2)
déc. 11 22:46:06 mde-oxalide systemd[1]: systemd-logind.service: Added fd 166 (session-c2) to fd store.
```

And when I unplug it:
```
déc. 11 22:46:15 mde-oxalide kernel: usb 1-1: USB disconnect, device number 11
déc. 11 22:46:15 mde-oxalide systemd[1]: systemd-logind.service: Received EPOLLERR on stored fd 166 (session-c2), closing.
déc. 11 22:46:15 mde-oxalide upowerd[698]: unhandled action 'unbind' on /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/0003:1050:0010.0021
déc. 11 22:46:15 mde-oxalide upowerd[698]: unhandled action 'unbind' on /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0
déc. 11 22:46:15 mde-oxalide upowerd[698]: unhandled action 'unbind' on /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1
```

Before taking those log I issue `sudo systemd-analyze set-log-level debug` (and checked which get-log-level).

```
dev-bus-usb-001-011.device: Changed dead -> plugged
```
seems strange…
Device unplugged:
```
max@mde-oxalide % systemctl status dev-bus-usb-001-011.device
● dev-bus-usb-001-011.device - Yubikey (v1 or v2)
   Follow: unit currently follows state of sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 22:46:06 CET; 8min ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1

déc. 11 22:46:06 mde-oxalide systemd[1]: dev-bus-usb-001-011.device: Changed dead -> plugged

max@mde-oxalide % systemctl status sys-devices-pci0000:00-0000:00:14.0-usb1-1\\x2d1.device
● sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device - Yubikey (v1 or v2)
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 18:54:11 CET; 4h 0min ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1
```

It seems a new dev-bus-usb-001-0X.device is created each time I plug my yubikey (X is incremented).
And stays active even if I unplug:
```
max@mde-oxalide % systemctl status dev-bus-usb-001-01{0..5}.device
● dev-bus-usb-001-010.device - Yubikey (v1 or v2)
   Follow: unit currently follows state of sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 22:44:38 CET; 16min ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1

déc. 11 22:44:38 mde-oxalide systemd[1]: dev-bus-usb-001-010.device: Changed dead -> plugged

● dev-bus-usb-001-011.device - Yubikey (v1 or v2)
   Follow: unit currently follows state of sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 22:46:06 CET; 14min ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1

déc. 11 22:46:06 mde-oxalide systemd[1]: dev-bus-usb-001-011.device: Changed dead -> plugged

● dev-bus-usb-001-012.device - Yubikey (v1 or v2)
   Follow: unit currently follows state of sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 22:52:48 CET; 7min ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1

● dev-bus-usb-001-013.device - Yubikey (v1 or v2)
   Follow: unit currently follows state of sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 22:54:16 CET; 6min ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1

● dev-bus-usb-001-014.device - Yubikey (v1 or v2)
   Follow: unit currently follows state of sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 22:56:32 CET; 4min 14s ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1

déc. 11 22:56:32 mde-oxalide systemd[1]: dev-bus-usb-001-014.device: Changed dead -> plugged

● dev-bus-usb-001-015.device - Yubikey (v1 or v2)
   Follow: unit currently follows state of sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device
   Loaded: loaded
   Active: active (plugged) since Mon 2017-12-11 22:57:14 CET; 3min 32s ago
   Device: /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1

déc. 11 22:57:14 mde-oxalide systemd[1]: dev-bus-usb-001-015.device: Changed dead -> plugged
```

But they are not listed by list-units
```
max@mde-oxalide % systemctl list-units | grep dev-bus
max@mde-oxalide %
```

**Comment by poettering** [comment]
hmm, this is very puzzling. Next thing to reproduce this would be to attach gdb and figure out what is going on internally. Are you comfortable with using gdb?

if so, plug in the device, then invoke "gdb" as root, type "attach 1", then "b device.c:819" (that's at least the line in current git, i.e. the one right after the `streq(action, "remove")` if check; it might actually be worth trying all this with current git first, before trying to debug this). Then type "c", now gdb is running and waiting for the break point to hit. now pull out the device. Now the breakpoint should be hit. Now skip over the swap_process_device_remove() invocation with "n", then single-step through  device_update_found_by_sysfs(), unit it finishes. Finally enter "detach" to detach again from PID 1, so that the system continues working as before. then, paste the session output somewhere, so I can have a look at the control flow for you. possibly use "p found" and "p d->found" and suchlike to dump the contents of local variables where relevant.

yeah, unless you are used to gdb this might be a bit complicate, but i have no better idea how to debug this right now.

**Comment by tchernomax** [comment]
@poettering I made a test VM and build the [latest systemd version](https://github.com/systemd/systemd/commit/2de2abad62d4f23f7d9af0c3871f9955896d085c).
The problem still happen.

I didn't check with gdb yet ; I have to find how to build/install systemd with debug symbol.

**Comment by ioquatix** [comment]
I seem to be having a similar issue, but with an iPhone X using USB to provide internet access. Even after the phone is disconnected, it still shows up as "loaded active plugged". The only way to fix this is to reboot or use a different USB port.

**Comment by ioquatix** [comment]
Hmm, so I checked, and it seems pretty similar - same bind/unbind messages.

If I try explicitly running `systemctl stop sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d1.device` it eventually times out.

**Comment by ioquatix** [comment]
I wonder if this is somehow related to https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=883425

**Comment by arvidjaar** [comment]
I can easily reproduce it in QEMU VM. As I can only hot-add USB storage device, I added similar rule to set `systemd` tag for corresponding USB device. Stracing systemd on device removal shows the single netlink message being received - with `ACTION=remove` for block device. I do not see any message for top level device.

This is different from hot plugging device when I get three netlink events - one for USB device `add` and two for block device (`add` and `change`).

So systemd is not even aware that device went away. Of course those devices are no more present under /sys, so they are actually gone.

... OK I know what's going on. This is rather unexpected udev behavior which recreates device properties from scratch on any event - even on `change` or `remove`. So with rule as above `remove` event for USB device won't contain `systemd` tag and will be ignored by systemd. You are expected to set properties **AGAIN** on **every** event ... even though one sensibly expects that udev already knows that device was tagged for system when it appeared ... but that's not even possible (or easy) because on removal no device properties exist anymore, so you cannot key off something like PID or VID - device is gone, where should you look it up?

systemd employs big hammer by always adding "systemd" tag for block device (or using match on device name) on any event which is why it appeared to work so far ...

**Comment by buczek** [comment]
@tchernomax, are you using kernel >= 4.12 ? We experience a similar issues and I traced it down to a failure of systemd-udevd to keep the device properties and tags when receiving _bind_ and _unbind_ events from the kernel, same thing  @arvidjaar describes for _change_ events. Because the systemd-tag is lost, the "remove" doesn't reach pid1.

You can easily check that by using

    udevadm monitor -t systemd

Do you see KERNEL and UDEV events for _add_ but only KERNEL events for _remove_ ? Do you see events of other typed between _add_ and _remove_ ? You can also check the suspected loss of properties by systemd-udevd with

    udevadm monitor -p

If so, the problem could be described as a manifestation of #8221.

**Comment by arvidjaar** [comment]
> Do you see KERNEL and UDEV events for add but only KERNEL events for remove ?

You misunderstand the problem. kernel and udev events are sent, but systemd explicitly filters out any event that does not contain `systemd` tag. If you match on specific device properties there is no way to add tag on `remove` event because

1. device is not more present so you cannot query properties
2. there is no way to look up "earlier" device properties added on previous `add` events in udev rule

So no way to create rule that matches specific device on `remove`.

This is limitation of how udev works. It is not new, in the past there was issue with interfaces losing properties after rename. It was solved in usual special-casing style by explicitly copying properties from database to "new" device ... for `move` action only.

**Comment by buczek** [comment]
> there is no way to look up "earlier" device properties added on previous add events in udev rule

Hmmm.  I thought the idea is, that `systemd-udevd` reads back the device properties from its `/var/run/udev/data` database when it receives the `remove` event from the kernel. This would include any tags set during the `add` event. [1] 

For our case,  I traced that `systemd-udev` indeed reads back the correct file from `/var/run/udev/data` during the `remove` event and I verified, that the file included all properties and the `sytemd`-tag after the `add` event, but it was trashed to emptiness during the `bind` and `unbind` events.  

I assumed, that `systemd-udevd` would otherwise generate a UDEV  `remove` event with all tags and properties restored, which would be received by `systemd` listening for UDEV events with the `systemd` tag and trigger the deactivation of its device unit.

[1] https://github.com/systemd/systemd/blob/5c69b31c137ffa24bca5a4628f9b4879d0f4e84a/src/udev/udev-event.c#L858

**Comment by arvidjaar** [comment]
> systemd-udevd reads back the device properties from its /var/run/udev/data database when it receives the remove event from the kernel

Yes, you are right here, my bad.

> For our case, I traced that systemd-udev indeed reads back the correct file from /var/run/udev/data during the remove event and I verified, that the file included all properties and the sytemd-tag after the add event, but it was trashed to emptiness during the bind and unbind events.

Yes, that remains true - on each non-`remove` event udev builds device properties from scratch. It is apparently possible to import existing properties using `IMPORT{db}`; but this applies to properties only, not to tags, which is the problem here.

So **every** rule must be updated to take into account any new kernel event type. 

Regarding original problem - I actually could workaround it using

`SUBSYSTEM=="usb", DEVTYPE=="usb_device", PRODUCT=="x/y/z", TAG+="systemd"`

Both `PRODUCT` and `DEVTYPE` are sent by kernel and so their existence does not depend on udev processing.

**Comment by buczek** [comment]
> each non-remove event udev builds device properties from scratch.

Yes. And I wonder, if this is intended. Users (e.g. pid1 here) rely on the fact that a tags added during the `add` event is stored in the udev database and is still available on the remove even, so that a `remove` is sent to monitor subscribers with this tag. You might be able to counteract that by modifying the rule files,  but shouldn't stored attributes and tags be persistent anyway? I'd  expect that.

**Comment by buczek** [comment]
@kaysievers, do you have an opinion?

**Comment by arvidjaar** [comment]
> And I wonder, if this is intended.

Probably yes. The only non-`remove` event for a long time was `change` and for `change` repeating the whole detection logic is actually most simple implementation (you need to query everything to find out was changed anyway at which point you can simply replace old data with new). In principle,  `bind` can follow the suite - device is there so we can simply repeat the same detection steps to collect properties again (and properties may change as device gets new driver as far I as I understand). But it simply does not work with `unbind`, because when event comes device is most like already gone so you can **not** query it to recreate database. 

Oh, and there are also `online` and `offline` events ... I wonder when they are emitted.

So I guess at least `unbind` should be treated similar to `remove` by default and properties should be initialized from database when rules are processed.

Anyway this is not the first report: https://github.com/systemd/systemd/issues/8221

**Comment by JimBobKennedy** [comment]
Any notion of when this will be fixed? It's apparently more widespread than just yubikey. It keeps my USB printer from working, too.

**Comment by poettering** [comment]
Duplicate of #8221.

**Comment by poettering** [comment]
(So, it kinda sucks that the kernel broke this without letting us know. So much about not breaking userspace...)

Anyway, I must admit I don't grok what "bind" and "unbind" precisely are supposed to do, the kernel commit msg I find more confusing than enlightening... i.e. it's not clear what we are supposed to do when we see this, and why these are events distinct from "change" anyway?

It appears to me that we should do the following at least though:

1. when we get unknown event types, ignore them
2. treat "bind" exactly like "change" or "add"
3. treat "unbind" like "remove" (maybe with the difference that we don't drop the database entry for the device, but probably not, an unbound device should probably not retain out of data, maybe instead programs should instead drop all devices that see "unbind".)

This under the assumption that "unbind" makes the device unavailable? (does it though? is this anywhere documented? kernel devs are so shitty at communicating what this is supposed to do, or even tell the main userspace consumer of this, gah). And "bind" just changes something that already has been there?

I think it would be nice to fix that for v240, but I don't think this should block the release. I will reassign the milestone to v241 hence. Happy to take a patch to fix this though (@arvidjaar, @yuwata?) , but i figure some research needs to be done first to figure out what these two actions precisely mean.

**Comment by poettering** [comment]
So, current udev git apparently already ignores events it doesn't know. Which is good.

**Comment by fbuihuu** [comment]
> So, current udev git apparently already ignores events it doesn't know. Which is good.

Hmm not sure (but also not sure what you meant exactly): AFAIR for unknown events udev treats them like "add" or "change" events and therefore udev rules are applied to "bind" events like it's done for any other "adding type" events.

And that's precisely what broke some udev rules out there because they explicitly match against "add" events and ignore "change" and "bind". However by ignoring "bind" (which can follow "add"), some properties or symlinks set during "add" can simply be canceled.

**Comment by poettering** [comment]
> Hmm not sure (but also not sure what you meant exactly): AFAIR for unknown events udev treats them like "add" or "change" events and therefore udev rules are applied to "bind" events like it's done for any other "adding type" events.

"bind" is a known event currently. But any events that the kernel will add in future is not known and with the current code will be ignored then until explicitly also added to the codebase.

Right now besides that "bind" and "unbind" are known by the codebase not much special treatment is done, and it is thus handled the same way as "change" in pretty much all places. In fact all events that are not "remove" (and to some degree "move") are treated exactly the same everywhere.

**Comment by poettering** [comment]
> And that's precisely what broke some udev rules out there because they explicitly match against "add" events and ignore "change" and "bind". However by ignoring "bind" (which can follow "add"), some properties or symlinks set during "add" can simply be canceled.

at least in our codebase (i.e. systemd, logind, ...) we only consider "remove" special, and everything else is treated the same. Similar in the rules file.

**Comment by fbuihuu** [comment]
That can work even though it's hard to tell if the new behavior can break existing udev rules.

Also if we keep using `ACTION=="remove",GOTO="xyz_end"`, we should ensure that any symlinks created on positive events will be removed on "unbind".

**Comment by poettering** [comment]
> That can work even though it's hard to tell if the new behavior can break existing udev rules.

Well, the status quo is borked, too. And people need to update their rules files anyway due to the kernel compat borkage, hence I am not too concerned I must say.

**Comment by sourcejedi** [comment]
> [other thread] Hmm, ok so this means that at the moment of the "unbind" the device becomes just a "skeleton" if you so will? It's still there, and it is still a device, but it's not really usable and has no useful properties?

> 2\. In apps treat "remove" and "unbind" as equivalent: consider the device as "unusable" from the moment that event is seen.

Nope.  On "unbind", you probably still have all the same properties.  After all, the physical device e.g. the PCI card is unchanged.  E.g. the network interface `eth0` will be removed entirely, but that is only a child device.

Apparently, the properties are available even if the "unbind" is emitted during removal!  https://github.com/systemd/systemd/issues/7587#issuecomment-381428545

But, when the device is being removed, you can't read any of the *attributes*.  And of course you can't do anything else to query the device.  And unfortunately "unbind" is being emitted during removal.

We don't tend to rely on the properties.  I don't know the reason off-hand, but I seem to remember the properties are not sufficient ID in every case.  (I want to say USB sub-something or endpoint or interface ?)

You'd think the kernel could have avoided this by suppressing "unbind" when it is really part of "remove"...  but actually no.  That made me think of working around it by checking if the device is still there when we process the event.  But that would be vulnerable to races - *and we were already vulnerable to this race*.  Right?  Imagine a `change` event happens, but then the device is removed before we can process the event.

The design that `change` events cause udev to regenerate the ID tags used for filtering (by libudev apps), suffers from a race condition v.s. removal...

> 3\. In udevd itself: make udev's TAGS "eternal"... the props database would always reflect the current state, which is a good thing. (the tags database would not, but that's fine, as its only purpose is matching anyway)

Yes, that sounds very promising.  It should solves the pre-4.12 race as well.

Maybe there's some weird device that changes types.  But I don't know any better way to handle that, and it sounds like such a device would be liable to cause more problems than the one we have here.

> This way all apps need to be patched to react to "unbind" correctly

If that's just your point 2\. above, I think we don't need it.  Maybe this does not suck too much :-).

**Comment by sourcejedi** [comment]
It might help people's mental models to know that you can bind and unbind drivers manually.  So "unbind" is *not* a one-way trip to "remove".

https://lwn.net/Articles/143397/ "Manual driver binding and unbinding" by Greg, in 2005.

**Comment by poettering** [comment]
> It might help people's mental models to know that you can bind and unbind drivers manually, by writing sysfs files. So "unbind" is _not_ a one-way trip to "remove".

Sure. But again: I am pretty sure apps should treat "unbind" like "remove" as long as they treat "bind" as equivalent to "add" or "change".

Or in other words app should just consider two categories:

1. events that make devices useful: "add", "change", "bind"
2. events that make devices useless: "remove", "unbind"

And ignore all other differences between the events. The above does *not* mean that the second type of event would really make the device go away and it could never recover, it just means "the device is not interesting to you anymore for now hence forget about it". if the first type of event is seen again, the device might become useful again.

And this also means that a device being temporarily unbound just means it becomes temporarily useless, and then when the "bind" is seen again it's useful again.  Which is OK.

**Comment by mbiebl** [comment]
If bind/unbind events are handled exactly like add/remove, what's the point in having separate events in the first place?

**Comment by ioquatix** [comment]
I thought about this and I think it's the difference between `umount` and `eject`. There are some tools where the difference is relevant.

**Comment by poettering** [comment]
Hmm, btw, anyone knows whether there's a safe and reliable way to detect whether a device is currently unbound? I have the suspicion if we consider an "unbind" event as reason to stop using a device, then during enumeration we'd need a way to figure out whether a device is in this post-"unbind" state (or even in its initial pre-"bind" state if it's that kind of device), to correctly ignore the device then too. Anyone has an idea?

**Comment by poettering** [comment]
> If bind/unbind events are handled exactly like add/remove, what's the point in having separate events in the first place?

Well, ask the kernel guys…

**Comment by poettering** [comment]
(i figure if we make tags eternal we also have to teach people to do "udevadm trigger -c remove ; udevadm trigger -c add" as the only safe way to fully make udev forget everything about a device, i.e. to flush out tags)

**Comment by sourcejedi** [comment]
If no driver is bound to a device, then the device has no `driver` link.

**Comment by keszybz** [comment]
The kernel commit I found is https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=1af824f085e813e06548212b18bbc788b16f60e6. Is there anything more verbose to look at?

**Comment by sourcejedi** [comment]
@keszybz the specific commit - not the merge commit - is: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=1455cf8dbfd06aa7651dcfccbadb7a093944ca65

My understanding of the reasons for the change, comes from the discussion on v1 (not v2) of the patch series that I linked to: https://lore.kernel.org/patchwork/patch/759829/

And I just realized Dmitry is currently working for Google, so he was probably doing work on Chromebooks.  E.g. that might explain why he noticed that systemd-udev was ignoring unknown events, but pre-systemd udev was not.  And maybe my confusion about the userspace firmware loading as well.

On preview: crossed posts :-).

**Comment by sourcejedi** [comment]
Which is very easy to do!

There are no implications for `udevd` itself.  Just keep using `ACTION!="remove"` / `ACTION=="remove"` + `GOTO=` in the rules.

The problem is in the details of how libudev clients filter on tags, and assume tags are never dropped from a device.  Maybe this was overlooked when tags were added to udevd?  I'm not sure what the story is there.

I think @poettering said we should _continue_ support dropping tags from a device e.g. for reconfiguring [uaccess] and [power-switch] tags - but it is wrong to think that we support this, at least for these two.  If you look at the systemd-logind code, it will not notice when tags are dropped from an existing device!    uaccess will fail to flush the ACL from the device, and logind will keep listening to the button unless you restart logind.

**Comment by sourcejedi** [comment]
> So, not sure what we can do. Ignore this and go ahead with making udev tags eternal? ideas?

I think we can fix it by removing the race condition in udev:

Before we evaluate the rules for a non-"remove" event, open the /sys path of the device.  After we've finished evaluating the rules, check if the syspath has been removed (`st_nlink == 0`).  If so, then do not save the updated udev device data (env DB, tags).  Just abort the processing before we save the data.  (It's probably also a good practice not to run the collected RUN programs, or broadcast the bad data to libudev apps.  Otherwise it's possible they could save the bad data and suffer the same problem we did).

Subsequently, we should see a `remove` event.  Hence we will process that `remove` event using the DB+tag data from the last "good" event.  We will broadcast the event to `libudev` clients, including the "good" data.

This doesn't fix the problem that happens when you remove udev tags.  It just means we won't be removing tags, when we lose the race against device removal.

**Comment by dtor** [comment]
Well, I've read the thread and, frankly, I am at loss for words. No, Lennart, the kernel developers did not break ABI, as existing events, as well as their attributes, were left untouched. However the uevent interface is supposed to be extendable and we have never committed to avoid adding new events or new attributes to existing events. Unfortunately systemd made questionable choice here and, and instead of only handling types of events that it knows how to handle, and ignoring and passing on to scripts/rules events that it is not aware of,  decided to handle everything that is not "remove" or "change" as "add". 

A few thoughts:

- treating unbind as remove is wrong. Unbind means that there is no driver that currently controls the device, but the device is currently present in the system and can be made operable if it is rebound to the same (or different) driver. That means that many of the device properties (such as modalias) are still available, and many of the tags are still valid.

- I think the only time you want to clear device tags/attributes is explicitly on "add" and "remove". Leave (pass on) events that you are not aware of. Maybe you need a syntax to remove a tag. 

- the reason that "bind" was added is because not all device attributes are created when device is registered. Quite often drivers add additional attributes when they start controlling the device. For example, many touch controller drivers will create "formware_id", "update_firrmware", etc attributes when driver binds to the physical device, which happens separately from i2c or spi device appearing in the system. If we want to use udev to check firmware version and opportunistically reflash it (this is separate from automatically loading firmware that is needed for device operation, i.e. here we are talking about controllers with NVRAM), then we can't do it at "add" time, we need to wait until driver is bound.Thus the new "bind" event.

- "add" means a new device appeared in the system. It may or may not be fully functional. "bind" means a driver was bound to a device and the device is now fully functional and all driver-specific attributes have been created. "unbind" means link between a driver and a device has been severed and the device is no longer fully functional, but may still be present in the system in which case it may be bound back to the same or different driver. "remove" means device is completely gone from the system (either because it was disconnected, or because the parent was unbound/removed).

- "be liberal with what you accept, be strict with what you emit". Explicitly handle events that you know how to handle, and do not do absolutely anything but pass on events that you are not aware of yet.

Thanks,
Dmitry

**Comment by poettering** [comment]
> Unfortunately systemd made questionable choice here and, and instead of only handling types of events that it knows how to handle, and ignoring and passing on to scripts/rules events that it is not aware of, decided to handle everything that is not "remove" or "change" as "add".

Well, iirc that behaviour was inherited all the way down when some kernel dev maintained udevd (gregkh). Also, what you are asking for there doesn't really work: udevd either can entirely suppress an event (i.e. ignore it entirely, not pass it on, so that noone sees it ever), or it can fully process an event (i.e. augment it with the various properties how it generally does stuff). You are asking for something in between: pass it on, but not process/augment it. But this would seriously confuse all apps, since suddenly all the props vanish from the device, and anyone seeing the event will assume that the new empty list of props are what is in effect.

Hence no, we only have two options: drop the event entirely (which then raises the question: why does the kernel generate this at all) or pass it on *fully processed*.

I am sorry, but at this point people have written lots of rules files and lots of apps consuming these events. And they generally wrote them in a style that "add", "remove" and "change" are the only events that appear. You appear to be of the strong opinion that that is wrong, and they should never have assumed that, but that doesn't really matter there, because reality is how reality is and not how you'd like it to be. I tell you as a matter of fact that clients generally don't expect anything besides to "add", "change" and "remove" to be seen. How I know this? I can look at the installed rules files on my system — and they all have the `ACTION!="add|change",GOTO="xyz_end"` line at the top which indicates they only consider "add" and "change" 'positive' actions and assume everything else invalidates the device. i.e. "bind" means all props and all tags vanish. The only rules files on my system that get this action matching right is actually the ones systemd ships itself. 

here's a quick list of packages installing rules files that will invalidate all tags/props on "bind" that are installed on my system here:

media-player-info-23-3.fc29.noarch
libwacom-0.32-1.fc29.x86_64
ykpers-1.19.0-3.fc29.x86_64
ModemManager-1.8.0-4.fc29.x86_64
libinput-1.12.2-1.fc29.x86_64
NetworkManager-1.12.4-2.fc29.x86_64
libgpod-0.8.3-24.fc29.x86_64

And this is just the stuff I have installed. Pretty much *every* *single* *other* package that installs rules files gets this wrong too, since they copy/paste from each other.

So yes, let me say this clearly: by introducing "bind" and you essentially broke almost all 3rd party packages with udev rules files.

**Comment by poettering** [comment]
>     * treating unbind as remove is wrong. Unbind means that there is no driver that currently controls the device, but the device is currently present in the system and can be made operable if it is rebound to the same (or different) driver. That means that many of the device properties (such as modalias) are still available, and many of the tags are still valid.

This really depends on the specific program. For example in systemd/PID 1 we expose devices as ".device" units, which can then be used to pull in services and suchlike at the right time. For this usecase the assumption is generally that when a .device unit changes from "inactive" to "active" that the device is then fully usable. i.e. services pulled in that way want to just open() the device (or whatever else they want to do with it), and assume it works. But now for devices that come up in this "unbound" state (i.e. that are not usable after "add", but only become usable after "bind") cannot be handled properly. 

Let me ask explicitly: is there any way how I can determine wether a device I discover (either through enumeration or because I see an "add" event) is in the "bound" state or not? i.e. how can I generically know for a device popping up whether I need to wait for a "bind" before considering it usable? There are some rumors that checking for the existance of a "driver" symlink  would work, but I don#t see how this could work, as for example block devices come up fully usable but certainly carry no "driver" symlink… So, my question is: how is a uevent that reports the change from the "unbound" to the "bound" state and another that reports the change in the other way useful if we cannot determine what the current state even is for a device?

**Comment by poettering** [comment]
> * I think the only time you want to clear device tags/attributes is explicitly on "add" and "remove". Leave (pass on) events that you are not aware of. Maybe you need a syntax to remove a tag.

This is not how this works or ever has worked. You are basically saying we shall not invalidate the tags and properties database on "change" even, right? But this just means that the data in the props/tags database will be basically always out of date and not reflect what is currently in effect. So far the "change" event was considered an event from the kernel to userspace asking it "please reprobe this device, things changed, please update your database, kthxby". You are now advocating though that "change" is now supposed to mean "something happened, please tell everyone that is has, but no, please leave the database out of date with the old data". THat doesn't make sense, does it?

**Comment by poettering** [comment]
> * the reason that "bind" was added is because not all device attributes are created when device is registered. Quite often drivers add additional attributes when they start controlling the device. For example, many touch controller drivers will create "formware_id", "update_firrmware", etc attributes when driver binds to the physical device, which happens separately from i2c or spi device appearing in the system. If we want to use udev to check firmware version and opportunistically reflash it (this is separate from automatically loading firmware that is needed for device operation, i.e. here we are talking about controllers with NVRAM), then we can't do it at "add" time, we need to wait until driver is bound.Thus the new "bind" event.

Why wouldn't have a regular "change" event sufficed for this too?

**Comment by poettering** [comment]
So, back to trying to improve things in this issue. The idea of making tags "eternal" might be salvageable. Right now doing TAGS+= on a rules file has two effects: it adds the mentioned tag as tag, and it also adds it to the colon separated colon list in the TAGS property. The latter can be useful for us. Let's redefine this so that:

1. The tags themselves become "eternal"
2. But the TAGS property itself always reflects the current status quo, i.e. from that list tags can vanish again.

Then, clients could use the tags *concept* itself for matching devices, but then use the TAGS *property* for determining the state of it.

So here's what I'd propose now for dealing with this mess:

1. In udev: make the tags concept eternal, but report in the TAGS property the real status quo.
2. In pid1 and logind: match by the respective tags but before considering a device check the TAGS property
3. In our NEWS file, ask all 3rd party udev consuming apps to update their rules files to change the `ACTION!="add|change",GOTO="xyz_end"` idiom to "ACTION=="remove",GOTO="xyz_end"`, so that they don't invalidate the device on "bind"
4. In our NEWS file, ask all 3rd party udev consuming apps to update their code in two regards:
   1. When they see "remove" monitor events they should stop caring for a device, and in all other events they should consider the device potentially interesting to them
   2. When they use tags for matching devices they should continue to do so but only consider the device relevant if the TAGS property contains the tag they are interested in, enclosed in `:`. (i.e the same as what we do ourselves following item 2 above)

**Comment by gregkh** [comment]
On Sat, Dec 08, 2018 at 03:54:36AM -0800, Lennart Poettering wrote:
> So here's what I'd propose now for dealing with this mess:

We can revert the BIND kernel change.  Forcing userspace packages to fix
this up really isn't ok.  Making this a CHANGE event would solve all of
this, right?

thanks,

greg k-h

**Comment by poettering** [comment]
> We can revert the BIND kernel change. Forcing userspace packages to fix this up really isn't ok. Making this a CHANGE event would solve all of this, right?

It wouldn't solve it, no. But it would reduce what broke by the kernel change. The summary is like this:

1. You could replace "bind" and "unbind" uevents by "change" in the kernel. This would mean:
   1. 3rd party apps and admins would not have to update their rules files to ensure "bind" is handled like "change"
   2. udev would still have to be fixed though, so that tags become "eternal" (as I proposed above), so that matching still works correctly even if a device is temporarily in an "unbound" state.
   3. 3rd party apps would still have to to be updated, so that they not only install matches for their tags, but also check the TAGS property every time they see a device event. (i.e. to deal with the outfall of making the tags "eternal")

2. You could remove "bind" and "unbind" in the kernel altogether. (And instead issue "remove" followed by "add" when this happens or so.) In this case neither udev nor 3rd party apps would hav to change anything. This is the only solution I see that breaks no userspace at all, i.e. makes userspace as happy as before kernel 4.12

3. You could leave everything as it is right now, then the changes I think are necessary are what I wrote above. i.e. udev needs patching, all 3rd party rules files and all 3rd party apps would need patching, too.

From my perspective though, it's kinda too late already. Every kernel since 4.12 is already borked like this iirc. (Though as I understand bind/unbind have not been used by too many subsystems initially, and only slowly gained adopted across subsystems?). I personally wouldn't ask you to revert anything. All I'd like to ask the kernel devs, is to consider the effects with userspace first, understand how udev actually works before doing changes like this kernel space, and in future please not add new uevents if "change" would have been OK too. Moreover, it would be good if any API changes like this are thought to the end, i.e. adding events that report changes between two states but not providing an API to determine the *current* status is clearly *not* thought to the end...

**Comment by dtor** [comment]
> Making this a CHANGE event would solve all of this, right?

Not quite as then we would not know if attributes we are looking for are supposed to be there or not, and so we would not know if we supposed to complain in the logs or everything is well. BIND gives a distinct point where we believe the device is fully functional.

**Comment by dtor** [comment]
> > Unfortunately systemd made questionable choice here and, and instead of only handling types of events that it knows how to handle, and ignoring and passing on to scripts/rules events that it is not aware of, decided to handle everything that is not "remove" or "change" as "add".
> 
> Well, iirc that behaviour was inherited all the way down when some kernel dev maintained udevd (gregkh). Also, what you are asking for there doesn't really work: udevd either can entirely suppress an event (i.e. ignore it entirely, not pass it on, so that noone sees it ever), or it can fully process an event (i.e. augment it with the various properties how it generally does stuff). You are asking for something in between: pass it on, but not process/augment it. But this would seriously confuse all apps, since suddenly all the props vanish from the device, and anyone seeing the event will assume that the new empty list of props are what is in effect.

Why would anyone assume anything? If they do not know about event they would simply ignore it. Why would they want to look at an event that they know nothing about?

> 
> Hence no, we only have two options: drop the event entirely (which then raises the question: why does the kernel generate this at all)

You seem to be under impression that all events emitted by the kernel are supposed to be for benefit of systemd exclusively.

Anyway, I tried explaining that instantiating a device is a multi-step process. There is a point where device gets enumerated by the system, and there is a point where it becomes fully functional. These are distinct points and significant time can pass between the two. For example, on ADD we attempt to load matching kernel modules.  

> or pass it on _fully processed_.

What is wrong with just giving it to users without changing systemd internal state if systemd does not understand what to actually make out of it?

> 
> I am sorry, but at this point people have written lots of rules files and lots of apps consuming these events. And they generally wrote them in a style that "add", "remove" and "change" are the only events that appear. You appear to be of the strong opinion that that is wrong, and they should never have assumed that, but that doesn't really matter there, because reality is how reality is and not how you'd like it to be. I tell you as a matter of fact that clients generally don't expect anything besides to "add", "change" and "remove" to be seen. How I know this? I can look at the installed rules files on my system — and they all have the `ACTION!="add|change",GOTO="xyz_end"` line at the top which indicates they only consider "add" and "change" 'positive' actions and assume everything else invalidates the device. i.e. "bind" means all props and all tags vanish. The only rules files on my system that get this action matching right is actually the ones systemd ships itself.

Huh? What on earth are you talking about? The packages indeed do  `ACTION!="add|change",GOTO="xyz_end"` and thus simply ignore the events they are currently not interested in. Such events as ONLINE and OFFLINE, I've seen references to MOVE, and also BIND and UNBIND. Where do you see them dropping tags and proprs from anything I do not know.

> 
> here's a quick list of packages installing rules files that will invalidate all tags/props on "bind" that are installed on my system here:
> 
> media-player-info-23-3.fc29.noarch

Does the right thing: ACTION!="add|change" GOTO end

> libwacom-0.32-1.fc29.x86_64

Does the right thing.

> ykpers-1.19.0-3.fc29.x86_64

Not checked

> ModemManager-1.8.0-4.fc29.x86_64

Seems to have explicit "bind" handling.

> libinput-1.12.2-1.fc29.x86_64

DTRT

> NetworkManager-1.12.4-2.fc29.x86_64

DTRT

> libgpod-0.8.3-24.fc29.x86_64

I do not see udev rules in main sources so don't know.

> 
> And this is just the stuff I have installed. Pretty much _every_ _single_ _other_ package that installs rules files gets this wrong too, since they copy/paste from each other.
> 
> So yes, let me say this clearly: by introducing "bind" and you essentially broke almost all 3rd party packages with udev rules files.

I beg to differ. The 3rd party packages are quite fine and are doing the right thing.

Any way we can move this over to email? I can't really compose a decent reply if all I have a window big enough for 4 lines of text in this github interface.

**Comment by dtor** [comment]
> 3. In our NEWS file, ask all 3rd party udev consuming apps to update their rules files to change the `ACTION!="add|change",GOTO="xyz_end"` idiom to `"ACTION=="remove",GOTO="xyz_end"`, so that they don't invalidate the device on "bind"

No, this request would indeed break them. `ACTION!="add|change",GOTO="xyz_end"` is indeed the right stanza for rules, unless they want to handle BIND explicitly and add some tags at that phase or perform some additional action. The tags/attributes/etc accumulate when rules firing, so as soon as systemd would stop resetting device state when receiving events it is not aware of, everything should start work properly.

**Comment by poettering** [comment]
> You seem to be under impression that all events emitted by the kernel are supposed to be for benefit of systemd exclusively.

I sense you want to turn this into a systemd-is-stupid thing. This is unfriendly and very much misguided. I'd appreciate it if you wouldn't go down this road. Thank you.

So, let me clear a few things up: "udevd" is one daemon shipped as part of the systemd package. "systemd" itself is another daemon shipped as part of the systemd package. "udev" processes, augments and passes on all uevents to various consumers. One of these consumers is "systemd". There are many other consumers. Your "bind" event broke *all* of them. And this is not systemd's fault, it's not even udev's fault really, you broke compat there with all udev versions and all consumers of udev events ever released, including the versions long before those evil dumb systemd people merged into into the same repo as systemd.

**Comment by poettering** [comment]
> What is wrong with just giving it to users without changing systemd internal state if systemd does not understand what to actually make out of it?

Please read up how uevents are processed by uevents. Here's what udev does:

When udev gets an uevent from the kernel, it applies a couple of "rules" on them, that augment what the kernel passed it. For example it runs "blkid" on block devices that read file system metadata. Or it runs "usb_id" which collects some USB metadata from sysfs, classifies it, and ultimately creates stable symlinks and does access management for it. These rules can do various things:

1. Add device node symlinks
2. Control access mode, ownership, ACLs on device nodes
2. Add additional properties to the device (i.e. extend what the kernel passed in anyway)
3. Add so-called "tags" to the devices, which allow filtering of uevents later on
4. and millions of other things.

After it completed all that, it then passes the uevents on to its clients. These uevents are heavily augmented now, with all the bits and pieces added by the udev rules. Besides passing these augmented uevents on it also stores the last seen properties in some database in /run, which can be used by clients which want to enumerate things, so see what the most recent data about devices is. The database on disk generally reflects the data acquired on the last uevent, the assumption is that whenever there's reason for the data to change the kernel would send a uevent. Whenever a uevent is seen any old data for that device is flushed out and replaced by the data derived from the new uevent.

Now, clients are usually only interested in a small subset of the devices. A storage manager only cares about block devices, and possibly even a subset of those (for example: no cd drvies, no floppies, …) A sound server only cares about sound devices, a display manager only about graphics devices. Thus they all apply filters. These filters are applied when enumerating (i.e. when such a program starts up and needs to know what is there), and when monitoring (i.e. when already running to see what new devices show up and which old devices go away). These filters are generally built on properties and tags (i.e. the stuff that rules files added). 

Now, you added the "bind" and "unbind" events. Most rules files use these `ACTION=="add|change",GOTO="xyz_end"` at the top, so that the rules applied are only executed on "add" and "change". This means that all the properties and tags they add to the device are only added in case of "add" and "change", and in all other events no properties and no tags are generated. And this again means the uevent passed on and the data now replacing the old database entry in /run will be stripped bare, it will basically be identical to the original kernel uevent data udev received originally. Now, this in turn means that any app that filters for specific properties and tags (as explained above), will not recv the uevent anymore (in case it monitors) or consider the device anymore during enumeration (because there are no properties/tags on it anymore), as both are now filtered away. And that's what breaks everything: devices disappear on "bind" from all apps view, because the filtering will filter them away as the events and database after bind basically lack everything those apps filtered on!

Is that understandable?

And this behaviour is no crazy invention of the stupid systemd people. It's how udev *always* worked really. It's what your fellow kernel developer greg (together with kay) came up with. So please, redirect and condescending comments on how stupid udev handles this (even though I think it does not), or how stupid apps handle this (even though i really don't think they do) towards your own kind kernel devs please. And please when touching the uevent code try to understand first how the consumers use it, thank you!

**Comment by poettering** [comment]
> Any way we can move this over to email? I can't really compose a decent reply if all I have a window big enough for 4 lines of text in this github interface.

You can resize the input field. There's a grip for that on the lower right of the edit box. Most web browsers have that.

That said, you can simply reply to the emails github sends you when somebody posts something here, it will then also show up here. In fact you can ignore the web interface entirely wen you want to participate in discussions on github.

**Comment by dtor** [comment]
On Sat, Dec 8, 2018 at 3:32 AM Lennart Poettering <notifications@github.com>
wrote:

> * treating unbind as remove is wrong. Unbind means that there is no driver that currently controls the device, but the device is currently present in the system and can be made operable if it is rebound to the same (or different) driver. That means that many of the device properties (such as modalias) are still available, and many of the tags are still valid.
>
>
> This really depends on the specific program. For example in systemd/PID 1
> we expose devices as ".device" units, which can then be used to pull in
> services and suchlike at the right time. For this usecase the assumption is
> generally that when a .device unit changes from "inactive" to "active" that
> the device is then fully usable. i.e. services pulled in that way want to
> just open() the device (or whatever else they want to do with it), and
> assume it works. But now for devices that come up in this "unbound" state
> (i.e. that are not usable after "add", but only become usable after "bind")
> cannot be handled properly.
>
> Let me ask explicitly: is there any way how I can determine wether a
> device I discover (either through enumeration or because I see an "add"
> event) is in the "bound" state or not? i.e. how can I generically know for
> a device popping up whether I need to wait for a "bind" before considering
> it usable? There are some rumors that checking for the existance of a
> "driver" symlink would work, but I don#t see how this could work, as for
> example block devices come up fully usable but certainly carry no "driver"
> symlink… So, my question is: how is a uevent that reports the change from
> the "unbound" to the "bound" state and another that reports the change in
> the other way useful if we cannot determine what the current state even is
> for a device?
>
Repeating what I said on another thread here: I do not think we have a
foolproof way to know whether a device requires/can have a driver or not. I
think we lost it when we unified hardware devices and class devices in
kernel :( Still, I think devices that are on physical buses do have
drivers, whereas virtual (or class devices) do not have one. Can you add a
lookup and see whether subsystem link points to a bus or class subtree for
a given device? Devices belonging to a class would not have a driver for
them and would be fully-functional as soon as they are created.

Another point: it is unlikely that the devices for which we emit "bind"
attributes are directly consumed by the upper layers. They are usually
foundation on which generic abstractions, such as block
device/partition/input device/serial port/etc are build on, and those are
usually the devices userspace is interested in. Nobody cares about i2c
touch controller itself, everyone want the input device that is child of
the controller as that is what represents useful to userspace abstraction.

Do you want a "device ready" event? I'm only half-joking here ;)

Thanks.

-- 
Dmitry

**Comment by boucman** [comment]
in the embedded world we do have i2c devices that don't have kernel drivers at all... arduino-like device we talk to directly with a custom upper-layer protocol...

**Comment by dtor** [comment]
On Sat, Dec 8, 2018 at 2:50 PM Boucman <notifications@github.com> wrote:

> in the embedded world we do have i2c devices that don't have kernel
> drivers at all... arduino-like device we talk to directly with a custom
> upper-layer protocol...
>
OK, right, but from the kernel POV in this case the device is not fully
functional: there nothing that properly handles the device in kernel. Even
if you are using systemd/udev you would continue to listening to the "add"
events and take it from there. Your use case should not be affected at all
by any of this.

Thanks.

-- 
Dmitry

**Comment by ioquatix** [comment]
As an end user, I find this discussion really interesting and I think everyone involved for explaining things clearly.

I get the feeling there are some frustrations regarding how people are interacting (people calling each other stupid or implying that). I can clearly see everyone here can make really amazing contributions to this discussion, and I know there are some areas where technically, some things have cross-wired either now or in the past. But I wish you all to not feel so much adversity towards each other, because when I read the discussion, I am in awe of everyone's technical ability and technical chops.

That being said, as a user, I can see both perspectives - the kernel's original model for events was too simplistic, and users (perhaps incorrectly) assumed that it would never change. Writing code like `(not event1|event2)` and assuming it would only ever trigger on `event3` is pretty fragile - that's an assumption I can see users have made which is causing some friction here. Additionally, I can see how users might look at the interface the kernel exposes, and expect that to be stable. For example, the change event could have been used to signal bind/unbind, or maybe `add` and `remove` could have extra tags to indicate if it was being bound or unbound, which might have avoided breaking user code.

That being said, I'm an advocate for clean logical interfaces, even if it means breaking backwards compatibility somewhat. I know it's a pain in the ass in the short term, but the long term benefits can be significant too (better logic, simpler logic, less bugs, easier to maintain, etc). So, even if there is some short term pain, I'd prefer that the kernel and systemd/udevd forge a way forward which makes the most sense w.r.t. the actual layers of code they have to deal with (i.e. ignoring borked user code). That's also taking into account, right now, it's already broken, so it can only get better :p

Also, looking at the `udev` rules, I wonder if some of these problems can't be alleviated for example, by explicitly dealing with `ACTION!="add|change"` as being equivalent to `ACTION=="remove"`.

Basically, I don't really know a huge amount about this, but I really wish everyone the best of luck to solve this problem, and I really wish everyone involved the love and respect they deserve. Thanks for all your effort!

**Comment by buczek** [comment]
> Also, looking at the udev rules, I wonder if some of these problems can't be alleviated for example, by explicitly dealing with ACTION!="add|change" as being equivalent to ACTION=="remove".

Alternative: Support legacy rules files, which are only prepared for "add","remove" and "change" event actions but allow rules files, which are able to handle all current and future event actions, too. E.g.  make rules files impotent (ignore assignments) if the event action is something else than "add", "remove" and "change" until it makes a special assignment and thereby declares that it is not legacy and can handle all event actions. Or maybe just decide by file extension.

**Comment by poettering** [comment]
> So I contend that this last part is the root of the issue. Instead of flushing the state and only storing attributes from the last uevent systemd (sorry, udev) should recognize the life cycle of devices in the kernel and accumulate the state across transitions. That means we should be collecting the attributes/tags from the initial ADD event, mix in additional tags from BIND event (if any) and finally flush the state on REMOVE. I can see where current CHANGE handling is proper as it refreshes the state. The refresh could b made explicit if you had a very early rule resetting all tags and attributes, then the core of udev would not need to bother.

Hmm, so tell me, I though I understood that bind/unbind was about attaching drivers to devices, and for example mode changing devices (e.g. 3g modems coming up as usb sticks that then change to be serial devices?) will first unbind the old driver and then rebind the new driver, right? So what you are proposing, i.e. that no props ever disappear would basically mean that all the props that where appropriate for a block device (such as metainfo about fs labels, symlinks...) would then later on be applied to a device that now is a modem? That sounds quite borked to me.

Or did I not get this right, and is bind/unbind not about changing drivers with an application in modechanging devices?

**Comment by poettering** [comment]
> We can revert the BIND kernel change. Forcing userspace packages to fix this up really isn't ok. Making this a CHANGE event would solve all of this, right?

@gregkh so thinking about this more, I think there's something the kernel *should* change, in order to improve compat with old udev: when a device is unplugged it currently sends "unbind" immediately followed by "remove" for it. The "unbind" will invalidate the per-device database in udev (old and new), and this means at the moment "remove" is seen the database is not there anymore, so that no tags/props can be read from it, which means most apps won't match the "remove" anymore (note that udev has special code to reuse the old per-device udev db on "remove", but only for "remove", to make matching easy, and this doesn't exist for "unbind", and can't really). If the kernel would not generate "unbind" in that case, and immediately go for the "remove" this compat breakage would be much less noticable. So, is there any reason for sending "unbind" right before "remove" on unplug? If not, would be great to suppress it in that case.

**Comment by poettering** [comment]
> Repeating what I said on another thread here: I do not think we have a foolproof way to know whether a device requires/can have a driver or not.

What about adding one? A new sysfs attr maybe called "bound". If it contains "1" the device is bound to a driver, and if it contains "0" it is not bound to a driver. If it doesn't exist, then the "bind"/"unbind" concept either doesn't apply to a device or is not known because kernel too old. And "bind" and "unbind" events would only be sent out by the kernel for devices that have this sysfs attr. In userspace, with that in place we could then add checks that if "bound" equals "0" we definitely ignore a device. And if it doesn't exist or is anything else we can#t really know and presume it is likely usable, but possibly not.

**Comment by dtor** [comment]
> > Repeating what I said on another thread here: I do not think we have a foolproof way to know whether a device requires/can have a driver or not.
> 
> What about adding one? A new sysfs attr maybe called "bound". If it contains "1" the device is bound to a driver, and if it contains "0" it is not bound to a driver. If it doesn't exist, then the "bind"/"unbind" concept either doesn't apply to a device or is not known because kernel too old. And "bind" and "unbind" events would only be sent out by the kernel for devices that have this sysfs attr. In userspace, with that in place we could then add checks that if "bound" equals "0" we definitely ignore a device. And if it doesn't exist or is anything else we can#t really know and presume it is likely usable, but possibly not.

The issue if that we can only set this attribute reliably to "1", which is really an equivalent of sending "bind". The devices that can be bound to a driver may at some point transition from "0" to "1", but we can not guarantee if and when it will happen. It could happen right away if driver is already loaded,it could happen in a second if udev manages to load the driver for us, it could happen in a few days when user installs an out-of-tree module, or it could never happen if driver is disabled, blacklisted, or simply does not exist at all.

I think looking at a subsystem is really the best option. Or simply continue what you were doing before, and assume the device is ready when it is added. The "bind" is really suited for tweaking driver-specific attributes, and once the driver is bound it usually produces yet another entity in which userspace is interested. Like I said, userspace is rarely interested in hardware devices for which we have "real" drivers, it usually looks for higher abstractions, like input devices, evdev, rtc, etc, etc that are expressed as "class" devices.

**Comment by sourcejedi** [comment]
@poettering your `usb_modeswitch` example does not apply, for two reasons.

I believe the kernel [only](https://lwn.net/Articles/651936/) re-reads the USB descriptors during a USB reset.  And when the kernel re-reads the USB descriptors:

> If the reset fails, <b>or the device's
> descriptors change from their values before the reset</b>, or the original
> configuration and altsettings cannot be restored, a flag will be set
> telling hub_wq to <b>pretend the device has been disconnected and then
> re-connected</b>.
>
> https://elixir.bootlin.com/linux/v4.19/source/drivers/usb/core/hub.c#L5533

This removes the USB device & child devices from sysfs (generating remove events). 

Secondly, it's true that udev generates properties for block devices, based on their content.  But when you unbind a driver which provides a block device, the block device is removed.  There is nothing you could call a skeleton block device.  As per @dtor's last paragraph, you do not need to worry about an `unbind` event on a block device.

```
$ udevadm monitor -k
monitor will print the received events for:
KERNEL - the kernel uevent

KERNEL[385.731078] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4 (usb)
KERNEL[385.731933] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0 (usb)
KERNEL[385.736178] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2 (scsi)
KERNEL[385.736217] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/scsi_host/host2 (scsi_host)
KERNEL[385.736258] bind     /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0 (usb)
KERNEL[385.736303] bind     /devices/pci0000:00/0000:00:14.0/usb2/2-4 (usb)
KERNEL[386.848237] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0 (scsi)
KERNEL[386.848298] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0 (scsi)
KERNEL[386.848331] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/scsi_disk/2:0:0:0 (scsi_disk)
KERNEL[386.848378] bind     /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0 (scsi)
KERNEL[386.848403] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/scsi_device/2:0:0:0 (scsi_device)
KERNEL[386.848684] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/scsi_generic/sg1 (scsi_generic)
KERNEL[386.848848] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/bsg/2:0:0:0 (bsg)
KERNEL[388.206129] add      /devices/virtual/bdi/8:16 (bdi)
KERNEL[388.229543] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/block/sdb (block)
KERNEL[388.229628] add      /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/block/sdb/sdb1 (block)
```

```
$ readlink -e /sys/devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/driver/
/sys/bus/scsi/drivers/sd
$ echo 2:0:0:0 | sudo tee /sys/bus/scsi/drivers/sd/unbind
```

```
KERNEL[936.897129] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/scsi_disk/2:0:0:0 (scsi_disk)
KERNEL[936.902203] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/block/sdb/sdb1 (block)
KERNEL[936.902247] remove   /devices/virtual/bdi/8:16 (bdi)
KERNEL[936.911645] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0/block/sdb (block)
KERNEL[936.911680] unbind   /devices/pci0000:00/0000:00:14.0/usb2/2-4/2-4:1.0/host2/target2:0:0/2:0:0:0 (scsi)
```

(I see the intermediate bus devices may or may not have drivers.  `/sys/bus/usb/devices/2-4/driver` points to the `usb` driver.  Unbinding the `usb` driver removes the child usb bus devices (and hence their children, including the block device).  But the devices `/sys/bus/scsi/{host2,target2:0:0}` do not have drivers.  So a bus device can have bus device children without having a driver bound to it.  It seems to me  those bus devices are fully operational despite the absence of a driver.  Maybe there's even more subtle details, but I really don't want to care about these details unless we can give a good reason.)

**Comment by dtor** [comment]
> > So I contend that this last part is the root of the issue. Instead of flushing the state and only storing attributes from the last uevent systemd (sorry, udev) should recognize the life cycle of devices in the kernel and accumulate the state across transitions. That means we should be collecting the attributes/tags from the initial ADD event, mix in additional tags from BIND event (if any) and finally flush the state on REMOVE. I can see where current CHANGE handling is proper as it refreshes the state. The refresh could b made explicit if you had a very early rule resetting all tags and attributes, then the core of udev would not need to bother.
> 
> Hmm, so tell me, I thought I understood that bind/unbind was about attaching drivers to devices, and for example mode changing devices (e.g. 3g modems coming up as usb sticks that then change to be serial devices?) will first unbind the old driver and then rebind the new driver, right? So what you are proposing, i.e. that no props ever disappear would basically mean that all the props that where appropriate for a block device (such as metainfo about fs labels, symlinks...) would then later on be applied to a device that now is a modem? That sounds quite borked to me.
> 
> Or did I not get this right, and is bind/unbind not about changing drivers with an application in modechanging devices?

No, IIRC the mode changing devices present themselves as a brand new device when they switch a mode, so we won't accumulate tags like that. However you do bring a valid point - in HID we can now automatically rebind from generic to a specialized driver, and thus we would need to undo whatever steps were done at "bind" time (assuming that there was something that needs undoing).

Note that this is not really unique issue with tags and not really limited to bind/unbind. Sometimes we need to do proper cleanup. For example, if I were to run a custom script on add, I would need to run a similar script on remove if the original script has side-effects extending beyond the device. So if "bind" does something we'll have to undo that at "unbind". We may provide some automatic facilities (I think in another post I asked whether we could mark tags and other attributes with action that added them, and automatically clear them when we receive complementing action), or we can simply tell users to clean up themselves (as long as we provide them with a way to remove a tag).

**Comment by poettering** [comment]
So implemented the "eternal" tags idea now, as noone could come up with anything smarter, and I think it's the best way ahead in a shitty situation. See PRs #11149 and #11138.

**Comment by poettering** [comment]
> Yes, there is, because it is done by different layers and code that does unbind does not necessarily knows the reason for unbinding. It would also not quite solve everything as the tags would still be wrong if you simply remove/unbind a driver.

That sounds like a kernel internal implementation detail to me. I am not sure why it's a good idea to let kernel implementation details leak to userspace like this.

**Comment by dtor** [comment]
> > Yes, there is, because it is done by different layers and code that does unbind does not necessarily knows the reason for unbinding. It would also not quite solve everything as the tags would still be wrong if you simply remove/unbind a driver.
> 
> That sounds like a kernel internal implementation detail to me. I am not sure why it's a good idea to let kernel implementation details leak to userspace like this.

It is kernel implementation detail. I explained though that what you are asking would not help solving the issue completely, so there is no point in trying to suppress unbind for one specific case.

**Comment by dtor** [comment]
It will not do that _reliably_. If driver gets unbound for some reason (unloaded, sysfs, etc) before device is removed, "unbind" will not be suppressed and you would still lose tags. The behavior would also be confusing and inconsistent, as we would sometimes send unbind and sometimes we would not.

**Comment by poettering** [comment]
Reassigning to v243, since we still have no idea what the right fix is.

**Comment by rleon** [comment]
I don't know if it should go to the separate issue and maybe PR https://github.com/systemd/systemd/pull/11138 was never merged, but we are experiencing the same behavior of zombie sysfs entries after "MOVE" event.

[leonro@server ~]$ udevadm --version
241

A real device, after rename:
  sys-devices-pci0000:00-0000:00:0a.0-infiniband-ibp0s10.device               loaded    active     plugged   MT27700 Family [ConnectX-4]                                       
An old device, before rename, but it is still visible in "systemctl -a":
  sys-devices-pci0000:00-0000:00:0a.0-infiniband-mlx5_0.device                loaded    active     plugged   MT27700 Family [ConnectX-4]  

Our device rename implementation on kernel boot:
https://github.com/linux-rdma/rdma-core/blob/master/kernel-boot/rdma-persistent-naming.rules

Any ideas?

Thanks

**Comment by kocjs** [comment]
So what is the suggested workaround?

Do we really have to restart the system after having used a USB device once? It looks like Windows is doing better!

**Comment by gdamjan** [comment]
> So what is the suggested workaround?
> 
> Do we really have to restart the system after having used a USB device once? It looks like Windows is doing better!

the workaround to match a remove event based on the `PRODUCT` and add the systemd tag works:
```
ACTION=="remove", SUBSYSTEM=="usb", ENV{PRODUCT}=="…", TAG+="systemd"
```
get the PRODUCT from `udevadm info <your device>`

**Comment by meltdown03** [comment]
I'm having this issue with kernel 5.5.7 and up. 5.5.6 works fine. The devices are FTDI FT2232HL and CP2102 usb-to-serial converters, though. Could my issue be related to the issue mentioned here?

**Comment by gregkh** [comment]
On Wed, Apr 01, 2020 at 07:05:32PM -0700, Neal MIller wrote:
> I'm having this issue with kernel 5.5.7 and up. 5.5.6 works fine. The
> devices are FTDI FT2232HL and CP2102 usb-to-serial converters, though.
> Could my issue be related to the issue mentioned here?

No idea.  Try asking on the linux-usb mailing list.

**Comment by jose1711** [comment]
Can confirm the issue on my system (ArchLinux, kernel 5.5.13)  and Wacom tablet. Workaround from https://github.com/systemd/systemd/issues/7587#issuecomment-381428545 fixes it for now.

**Comment by meltdown03** [comment]
> Can confirm the issue on my system (ArchLinux, kernel 5.5.13) and Wacom tablet. Workaround from [#7587 (comment)](https://github.com/systemd/systemd/issues/7587#issuecomment-381428545) fixes it for now.

That didn't work for my CP2102. Maybe I had the `.rules` filed edited incorrectly. I've seen a couple bug reports for the exact issue I am having:
[RedHat Bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1813530)
[Gentoo Forums](https://forums.gentoo.org/viewtopic-p-8441008.html#8441008)
